app/models/space.rb

```diff
@@ -101,41 +105,50 @@ def potential_duplicates
   # NOTE: this expects a scope for spaces but returns an array
   # preferably we would find some way to return a scope too
   def self.filter_on_facilities(spaces, filtered_facilities)
-    results = spaces.includes(:space_facilities).filter_map do |space|
-      relevant_space_facilities = space.relevant_space_facilities
-
-      # If no relevant matches at all, exclude the space:
-      next unless (filtered_facilities & relevant_space_facilities.map { |sf| sf.facility.id }).any?
-
-      space.score_by_filter_on_facilities(filtered_facilities, relevant_space_facilities)
-    end
+    results = spaces.includes(:space_facilities)
+                    .where(space_facilities: { relevant: true, facility_id: filtered_facilities })
+                    .filter_map(&:score_by_filter_on_facilities)
 
     results.sort_by(&:score).map(&:space)
   end
 
-  def score_by_filter_on_facilities(filtered_facilities, relevant_space_facilities)
-    score = 0
-    relevant_space_facilities.each do |space_facility|
-      next unless filtered_facilities.include?(space_facility.facility_id)
-
+  def score_by_filter_on_facilities
+    score = 0.0
+    space_facilities.each do |space_facility|
       # The more correct matches the lower the number.
       # this is so the sort_by later will be correct as it sorts by lowest first
       # we could do a reverse on the result of sort_by but this will incur
       # a performance overhead
-      if space_facility.likely?
-        score -= 2
-      elsif space_facility.maybe?
-        score -= 1
-      elsif space_facility.unlikely?
-        score += 1
-      elsif space_facility.impossible?
-        score += 2
-      end
+
+      score -= score_from_experience(space_facility)
+
+      # Add a score for the star_rating to use as a tie-breaker.
+      score -= score_from_star_rating
```

Comment:
This should not be done for each space_facility! Should instead be done only once, and outside this loop

Reply:
Fixed